fluidkeys/keyrotate.go

```diff
@@ -3,8 +3,200 @@ package main
+	"github.com/fluidkeys/fluidkeys/fingerprint"
+	"github.com/fluidkeys/fluidkeys/keytableprinter"
+	"github.com/fluidkeys/fluidkeys/pgpkey"
+	"github.com/fluidkeys/fluidkeys/status"
 )
 
+func keyRotate(dryRun bool) exitCode {
+	keys, err := loadPgpKeys()
+	if err != nil {
+		panic(err)
+	}
+
+	keytableprinter.Print(keys)
+
+	if dryRun {
+		return runKeyRotateDryRun(keys)
+	} else {
+		return runKeyRotate(keys)
+	}
+}
+
+func runKeyRotateDryRun(keys []pgpkey.PgpKey) exitCode {
+	fmt.Printf("\nFluidkeys will perform the following actions.\n")
+	fmt.Printf("\nYou'll be asked to confirm before making any changes.\n")
+
+	var keysWithActions []*pgpkey.PgpKey
+
+	for i := range keys {
+		key := &keys[i]
+		warnings := status.GetKeyWarnings(*key)
+		actions := status.MakeActionsFromWarnings(warnings)
+		printKeyHeaderAndActions(key, actions)
+
+		if len(actions) > 0 {
+			keysWithActions = append(keysWithActions, key)
+		}
+	}
+
+	printImportBackIntoGnupg(keysWithActions)
+
+	fmt.Printf("\nTo start, run `fk key rotate`\n")
+	return 0
+}
+
+func runKeyRotate(keys []pgpkey.PgpKey) exitCode {
+	fmt.Printf("\nFluidkeys will perform the following actions.\n")
+	fmt.Printf("\n%s\n", colour.Warn("Take time to review these actions."))
+
+	var anyKeysHadActions = false
+	var keysModifiedSuccessfully []*pgpkey.PgpKey
+	passwords := make(map[fingerprint.Fingerprint]string)
+
+	numErrorsEncountered := 0
+
+	for i := range keys {
+		key := &keys[i] // get a pointer here, not in the `for` expression
+		warnings := status.GetKeyWarnings(*key)
+		actions := status.MakeActionsFromWarnings(warnings)
+		printKeyHeaderAndActions(key, actions)
+
+		if len(actions) > 0 {
+			anyKeysHadActions = true
+		} else {
+			continue // nothing to do. next key.
+		}
+
+		if promptYesOrNo("    Run these actions? [Y/n] ", true) == false {
+			fmt.Printf("    OK, skipped.\n")
+			continue // next key
+		}
+
+		key, password, err := getDecryptedPrivateKeyAndPassword(key)
+
+		action := "Load private key from GnuPG" // TODO: factor into func
+		if err != nil {
+			printCheckboxFailure(action, err)
+			continue
+		} else {
+			printCheckboxSuccess(action)
+		}
+
+		err = runActions(key, actions)
+		if err != nil {
+			numErrorsEncountered += 1
+			fmt.Printf("\n    Skipping remaining actions for %s\n", displayName(key))
+			continue // Don't run any more actions
+		} else {
+			message := fmt.Sprintf("Successfully updated keys for %s", displayName(key))
+			fmt.Printf("\n    %s\n", colour.Info(message))
+			keysModifiedSuccessfully = append(keysModifiedSuccessfully, key)
+			passwords[key.Fingerprint()] = password
+		}
+	}
+
+	if !anyKeysHadActions {
+		fmt.Printf("\n%s\n", colour.Green("✔ All keys look good — nothing to do."))
+		return 0 // success! nothing to do
+	}
+
+	if !runImportBackIntoGnupg(keysModifiedSuccessfully, passwords) {
+		numErrorsEncountered += 1
+	}
+
+	if numErrorsEncountered > 0 {
+		message := fmt.Sprintf("%d error(s) while running rotate.", numErrorsEncountered)
```

Comment:
Could you use `humanize.Pluralize` here to make it a touch nicer in the user interface?

Reply:
Ah great plan